[CONTP-535] optimise tagger server chunking function using iter package

adel121 opened the pull request:
<!--
* Contributors are encouraged to read our [CONTRIBUTING](/CONTRIBUTING.md) documentation.
* Both Contributor and Reviewer Checklists are available at https://datadoghq.dev/datadog-agent/guidelines/contributing/#pull-requests.
* The pull request:
  * Should only fix one issue or add one feature at a time.
  * Must update the test suite for the relevant functionality.
  * Should pass all status checks before being reviewed or merged.
* Commit titles should be prefixed with general area of pull request's change.
* Please fill the below sections if possible with relevant information or links.
-->
### What does this PR do?

This PR leverages the `iter` golang package introduced in golang 1.23.0 in order to optimise the chunk function used in the tagger server. 

### Motivation

Optimise memory consumption of the chunking function used to split the tagger stream events into a sequence of chunks, each having a bounded size (in bytes).

### Describe how you validated your changes
<!--
Validate your changes before merge, ensuring that:
* Your PR is tested by static / unit / integrations / e2e tests
* Your PR description details which e2e tests cover your changes, if any
* The PR description contains details of how you validated your changes. If you validated changes manually and not through automated tests, add context on why automated tests did not fit your changes validation.

If you want additional validation by a second person, you can ask reviewers to do it. Describe how to set up an environment for manual tests in the PR description. Manual validation is expected to happen on every commit before merge.

Any manual validation step should then map to an automated test. Manual validation should not substitute automation, minus exceptions not supported by test tooling yet.
-->
Unit tests are implemented.
E2E should be enough since they validated the remote tagger is still working.

Benchmark Tests Results:
```
go test -bench . -benchmem 
goos: linux
goarch: arm64
pkg: github.com/DataDog/datadog-agent/comp/core/tagger/server
BenchmarkChunkLazyChunking-10                  6         189819611 ns/op        80000906 B/op   10000001 allocs/op
BenchmarkChunkSliceChunking-10                 3         480011264 ns/op        1281662296 B/op 10000051 allocs/op
PASS
ok      github.com/DataDog/datadog-agent/comp/core/tagger/server        4.351s
```

### Possible Drawbacks / Trade-offs
None

### Additional Notes
<!--
* Anything else we should know when reviewing?
* Include benchmarking information here whenever possible.
* Include info about alternatives that were considered and why the proposed
  version was chosen.
-->

Lazy chunking is enabled by default with the config flag `tagger_server.use_lazy_chunking.enabled`. 
This can be useful to disable the new implementation in case of regressions.
This can be removed later when we no longer need to keep the old implementation.

adel121 committed:
optimise tagger server chunking function using iter package

clamoriniere reviewed (commented):
The change looks good.
Could you try to put in place a benchmark to compare the 2 implementations?

adel121 commented:
@clamoriniere 

I updated the PR with benchmark tests and readme file

I also added an internal feature flag that we can use to disable the new implementation in case of regression.

Thanks

adel121 commented:
Closing in favour of [this](https://github.com/DataDog/datadog-agent/pull/31837) PR.

The [other](https://github.com/DataDog/datadog-agent/pull/31837) PR provides an approach that is even more optimised than using iterators.
